File: is_rosetta.m
function r = is_rosetta()

r = false;

if ~ismac
  return
end

% uname -m reports "x86_64" from within Matlab on Apple Silicon if using Rosetta

[ret, raw] = system("sysctl -n sysctl.proc_translated");
r = ret == 0 && startsWith(raw, "1");

end

File: is_rosetta.cpp
#include <iostream>

#ifdef __APPLE__
#include <sys/sysctl.h>
#endif

bool is_rosetta(){

#ifdef __APPLE__
   int ret = 0;
   size_t size = sizeof(ret);

   if (sysctlbyname("sysctl.proc_translated", &ret, &size, nullptr, 0) < 0)
   {
      if (errno == ENOENT)
         return false;
      throw;
   }

   return (ret == 1);
#else
   return false;
#endif
}

int main(int argc, char ** argv) {

  if (is_rosetta()) {
    std::cout << argv[0] << " is using Rosetta" << std::endl;
  }
  else {
    std::cout << argv[0] << " is not using Rosetta" << std::endl;
  }

  return EXIT_SUCCESS;
}
